In [2]:
import pandas as pd

Q1. Pandas version
What version of Pandas did you install? 2.3.1

In [3]:
pd.__version__

'2.3.1'

In [4]:
fuel = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')

In [5]:
fuel.head(5)

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

In [6]:
fuel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


Q2. Records count

How many records are in the dataset? 9704

Q3. Fuel types

How many fuel types are presented in the dataset? 2

In [10]:
fuel['fuel_type'].value_counts()

fuel_type
Gasoline    4898
Diesel      4806
Name: count, dtype: int64

Q4. Missing values
How many columns in the dataset have missing values?

4

Q5. Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?
23.75


In [11]:
asia_cars = fuel[fuel['origin'] == 'Asia']

asia_cars['fuel_efficiency_mpg'].max()

np.float64(23.759122836520497)

Q6. Median value of horsepower

1. Find the median value of the horsepower column in the dataset. - 149.0
1. Find the median value of the horsepower column in the dataset.. - 152.0
3. Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.
4. Now, calculate the median value of horsepower once again.


In [12]:
# Step 1. Find the median value of the horsepower column in the dataset
fuel['horsepower'].median()

np.float64(149.0)

In [13]:
# Step 2. Find the median value of the horsepower column in the dataset.
fuel['horsepower'].mode()

0    152.0
Name: horsepower, dtype: float64

In [14]:
fuel['horsepower'] = fuel['horsepower'].fillna(152)

In [15]:
fuel['horsepower'].mode()

0    152.0
Name: horsepower, dtype: float64

Has it changed?

Yes - it has changed

Q7. Sum of weights

1. Select all the cars from Asia
2. Select only columns vehicle_weight and model_year
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the sum of all the elements of the result? 0.5187709081074016


In [16]:
import numpy as np
# Step 2 : Select only columns vehicle_weight and model_year of cars from Asia
asia_weight_year = asia_cars[['vehicle_weight','model_year']]

# Step 3&4 : Get the underlying NumPy array. Let's call it X
X = asia_weight_year[:7].values

# Step 5: Compute X transpose times X
XTX = X.T.dot(X)

# Step 6: Invert XTX
XTX_inv = np.linalg.inv(XTX)

# Step 7 : Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y= [1100, 1300, 800, 900, 1000, 1100, 1200]

# Step 8 : Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
w = XTX_inv @ X.T @ y

# Step 9 : What's the sum of all the elements of the result?
np.sum(w)

np.float64(0.5187709081074016)